# GWAS-data-simulator
Simulating GWAS data in [PLINK](https://www.cog-genomics.org/plink/) format with GWAsimulator tool using Hapmap3 data

# Description
This repository presents methods to simulate GWAS data using GWAsimulator tool, it consists of:
- Use HapMap3 data as reference for GWAsimulator.
- Convert HapMap3 phased data in GWAsimulator input format.
- Adjust the generated data to [PLINK](https://www.cog-genomics.org/plink/) format (.ped and .map files).
- Generate admixed population data, this could be useful in population stratification studies.

Either this repository consider HapMap3 data as input reference, the algorithm could be used to other references data such as ```1000 Genome Project``` data.
 
11 populations were included in [HapMap3](https://www.sanger.ac.uk/resources/downloads/human/hapmap3.html), we will give an example of generating a data of two admixed populations for ```CEU``` (Utah residents with Northern and Western European ancestry from the CEPH collection) and ```YRI``` (Yoruba in Ibadan, Nigeria). You can change other populations or add more than 2 populations following the same procedure.

The choice of GWAsimulator input parameters (in control.dat file) will not be discussed in this repository, please check the [documentation](http://biostat.mc.vanderbilt.edu/wiki/pub/Main/GWAsimulator/GWAsimulator_v2.0.pdf) provided by the authors for details.

# What is GWAsimulator?
A C++ software that simulates genotype (SNP chips) for the whole human genome in case-control. Different populations could be generated by following the Linkage Disequilibrium patterns of a reference data, such as HapMap3. In practice, the algorithm takes as input:
- Reference data: phased data in 23 files (one file for each chromosome: from 1 to 23 for chromosome X), the alleles must decoded as ```0``` or ```1```. The program considers only  bi-allelic SNP.
- Control file: it includes information about the desired simulated data such as: output file format, disease locus position and the number of males and females generated samples...

# Requirements
- Python
- [GWAsimulator](http://biostat.mc.vanderbilt.edu/GWAsimulator.)
- [HapMap3 data](ftp://ftp.ncbi.nlm.nih.gov/hapmap/phasing/2009-02_phaseIII/HapMap3_r2)
- Pandas

# References:
Li C et al., GWAsimulator: a rapid whole-genome simulation program, Bioinformatics (2008).
